aten/src/ATen/native/cudnn/MHA.cpp

```diff
@@ -292,6 +292,21 @@ auto fixSizeOneDimStrideSDPA(
   }
   return strides;
 }
+
+void alloc_with_matching_layout(const Tensor& q, Tensor& output, const std::vector<int64_t>& shape) {
+  TORCH_CHECK(shape.size() == q.sizes().size(), "cuDNN SDPA alloc_with_matching_layout got requested shape ndim != q ndim");
+  // get the "fill order," which is just an argsort on the strides
+  std::vector<int> fill_order(shape.size());
+  std::iota(fill_order.begin(), fill_order.end(), 0);
+  std::stable_sort(fill_order.begin(), fill_order.end(), [&q](int idx1, int idx2) {return q.strides()[idx1] < q.strides()[idx2];});
+  std::vector<int64_t> ordered_strides(shape.size());
+  int64_t current_stride = 1;
+  for (const int dim_idx : fill_order) {
+    ordered_strides[dim_idx] = current_stride;
+    current_stride *= shape[dim_idx];
+  }
+  output = at::empty(at::IntArrayRef(shape), q.options()).as_strided(at::IntArrayRef(shape), at::IntArrayRef(ordered_strides), 0);
```

Comment:
Why one couldn't use `empty_like(q)`?

Reply:
Output is not guaranteed to have the same shape as query, (e.g., `d_q != d_v`)

Reply:
Additionally we are matching q's layout order but not really its exact strides or layout 1:1 as q can be in a chunked/packed layout where it's not contiguous given any permutation of dimensions.